packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js

```diff
@@ -0,0 +1,270 @@
+      this.logger.debug({
+        at: "ForexDailyPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "ForexDailyPriceFeed",
+      message: "Updating ForexDailyPriceFeed",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // Find the closest day that completed before the beginning of the lookback window, and use
+    // it as the start date.
+    const startLookbackWindow = currentTime - this.lookback;
+    const startDateString = this._secondToDateTime(startLookbackWindow);
+    const endDateString = this._secondToDateTime(currentTime);
+
+    // 1. Construct URL.
+    // See https://exchangeratesapi.io/ for how this url is constructed.
+    const url = [
+      "https://api.exchangeratesapi.io/history?",
+      `start_at=${startDateString}&end_at=${endDateString}`,
+      `&base=${this.base}&symbols=${this.symbol}`
+    ].join("");
+
+    // 2. Send request.
+    const historyResponse = await this.networker.getJson(url);
+
+    // 3. Check responses.
+    if (
+      !historyResponse?.rates ||
+      Object.keys(historyResponse.rates).length === 0 ||
+      Object.values(historyResponse.rates).some(rate => !rate[this.symbol])
+    ) {
+      throw new Error(`🚨Could not parse price result from url ${url}: ${JSON.stringify(historyResponse)}`);
+    }
+
+    // 4. Parse results.
+    // Return data structure:
+    // {
+    //   "rates": {
+    //     "2021-03-16": {"EUR":0.8385041087},
+    //     "2021-03-15": {"EUR":0.8389261745},
+    //   },
+    //     "start_at": "2021-03-15",
+    //     "end_at": "2021-03-16",
+    //     "base": "USD"
+    // }
+    const newHistoricalPricePeriods = Object.keys(historyResponse.rates)
+      .map(dateString => ({
+        // Output data should be a list of objects with only the open and close times and prices.
+        openTime: this._dateTimeToSecond(dateString) - 24 * 3600,
+        closeTime: this._dateTimeToSecond(dateString),
+        closePrice: this.convertPriceFeedDecimals(historyResponse.rates[dateString][this.symbol])
+      }))
+      .sort((a, b) => {
+        // Sorts the data such that the oldest elements come first.
+        return a.openTime - b.openTime;
+      });
+
+    // 5. Store results.
+    this.currentPrice = newHistoricalPricePeriods[newHistoricalPricePeriods.length - 1].closePrice;
+    this.historicalPricePeriods = newHistoricalPricePeriods;
+    this.lastUpdateTime = currentTime;
+  }
+
+  _secondToDateTime(inputSecond) {
+    return moment.unix(inputSecond).format("YYYY-MM-DD");
+  }
+  _dateTimeToSecond(inputDateTime) {
+    return moment(inputDateTime, "YYYY-MM-DD").unix();
+  }
```

Comment:
I'm not sure how precise we want to be around these seconds -> day conversions, but the ECB data is [published at 16:00 CET](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html) (computed at 14:15 CET, which might also be a good choice for stamping it), which is 3:00PM UTC and 11:AM EST. Using the terminology from this price feed, I would think the open time for the date string "2021-03-12" would be 2021-03-11 at 16:00 CET and the close time would be 2021-03-12 at 16:00 CET (or 14:15 if you prefer the compute time).

Thoughts?

Reply:
Great point, check out [this commit](https://github.com/UMAprotocol/protocol/pull/2718/commits/c7bf6695d06d7add6ff41fd3648cefa23e41831a)

Reply:
This is really really elegant. +1